Feat: AI sanitizer plugin (#8479)
Kong/docs.konghq.com: app/_hub/kong-inc/ai-sanitizer/overview/_index.md

---
nav_title: Overview
---

> The Data Sanitization plugin for Kong Gateway helps protect sensitive information in client request bodies before they reach upstream services.

Author: ```suggestion
The Data Sanitization plugin for {{site.base_gateway}} helps protect sensitive information in client request bodies before they reach upstream services.
```

By integrating with an external PII service, the plugin ensures compliance with data privacy regulations while preserving the usability of request data.
It supports multiple sanitization modes, including replacing sensitive information with fixed placeholders or generating synthetic replacements that retain category-specific characteristics.

> Additionally, it offers an optional restoration feature, allowing the original data to be reinstated in responses when needed.

Author: ```suggestion

Additionally, it offers an optional restoration feature, allowing the original data to be reinstated in responses when needed.
```

> 

Author: ```suggestion

{:.note}
> This plugin extends the functionality of the [AI Proxy plugin](/hub/kong-inc/ai-proxy/), and requires AI Proxy to be configured first. 
Check out the [AI Gateway quickstart](/gateway/latest/get-started/ai-gateway/) to get an AI proxy up and running within minutes!
```

## How it works

> 1. When a request reaches Kong Gateway, the plugin intercepts the request body and forwards it to the external PII service.

Author: ```suggestion
1. When a request reaches {{site.base_gateway}}, the plugin intercepts the request body and forwards it to the external PII service.
```

2. The PII service analyzes the content, identifies sensitive information, and applies the selected sanitization method (fixed placeholders or category-based synthetic replacements).

> 3. The sanitized request body is then forwarded to the upstream service through ai-proxy plugin.

Author: ```suggestion
3. The sanitized request body is then forwarded to the upstream service through the AI Proxy or AI Proxy Advanced plugin.
```

4. If the restoration feature is enabled, the plugin can restore the original data in responses before returning them to the client, ensuring a seamless user experience.

{:.note}

> > **Note**: This plugin requires the AI Proxy plugin to function properly, please make sure the AI Proxy plugin is enabled and configured correctly for this plugin to work as expected.

Author: ```suggestion
> **Note**: This plugin requires the AI Proxy plugin or the AI Proxy Advanced plugin to function properly. Make sure one of the proxy plugins is enabled and configured correctly for this plugin to work as expected.
```

---
nav_title: Overview
---

The AI Sanitizer plugin for {{site.base_gateway}} helps protect sensitive information in client request bodies before they reach upstream services.
By integrating with an external PII service, the plugin ensures compliance with data privacy regulations while preserving the usability of request data.
It supports multiple sanitization modes, including replacing sensitive information with fixed placeholders or generating synthetic replacements that retain category-specific characteristics.

Additionally, it offers an optional restoration feature, allowing the original data to be reinstated in responses when needed.

{:.note}
> This plugin extends the functionality of the [AI Proxy plugin](/hub/kong-inc/ai-proxy/) or the [AI Proxy Advanced](/hub/kong-inc/ai-proxy-advanced/) plugin, and requires an AI proxy to be configured first. 
Check out the [AI Gateway quickstart](/gateway/latest/get-started/ai-gateway/) to get an AI proxy up and running within minutes!

> The AI Sanitizer plugin relies on the AI PII Anonymizer Service for identifying and sanitizing sensitive information, which can be deployed as a standalone service or Docker container. Check out the [AI Sanitizer plugin's configuration options](/hub/kong-inc/ai-sanitizer) for more information on how to configure the plugin with the AI PII Anonymizer Service.

Author: ```suggestion
The AI Sanitizer plugin relies on the AI PII Anonymizer Service for identifying and sanitizing sensitive information, which can be deployed as a Docker container. Check out the [AI Sanitizer plugin's configuration options](/hub/kong-inc/ai-sanitizer) for more information on how to configure the plugin with the AI PII Anonymizer Service.
```

---
nav_title: Overview
---

The AI Sanitizer plugin for {{site.base_gateway}} helps protect sensitive information in client request bodies before they reach upstream services.
By integrating with an external PII service, the plugin ensures compliance with data privacy regulations while preserving the usability of request data.
It supports multiple sanitization modes, including replacing sensitive information with fixed placeholders or generating synthetic replacements that retain category-specific characteristics.

Additionally, it offers an optional restoration feature, allowing the original data to be reinstated in responses when needed.

{:.note}
> This plugin extends the functionality of the [AI Proxy plugin](/hub/kong-inc/ai-proxy/) or the [AI Proxy Advanced](/hub/kong-inc/ai-proxy-advanced/) plugin, and requires an AI proxy to be configured first. 
Check out the [AI Gateway quickstart](/gateway/latest/get-started/ai-gateway/) to get an AI proxy up and running within minutes!

{:.note}

> > The AI Sanitizer plugin was relying on the [AI Pii Services](https://github.com/Kong/ai-pii-service) service for its AI capabilities, which can be deployed as a standalone service or docker container. Check out the [AI Sanitizer plugin](/hub/kong-inc/ai-sanitizer) for more information on how to config the plugin with the AI Pii Services.

Reviewer: I am not sure if I can use the URL https://github.com/Kong/ai-pii-service, because we don't know if this repository will be open source.

> {:.note}
> > The AI Sanitizer plugin is relying on the [AI PII Services](https://github.com/Kong/ai-pii-service) service for identifying and sanitizing sensitive information, which can be deployed as a standalone service or docker container. Check out the [AI Sanitizer plugin](/hub/kong-inc/ai-sanitizer) for more information on how to config the plugin with the AI PII Services.

Author: ```suggestion
The AI Sanitizer plugin relies on the AI PII Anonymizer Service for identifying and sanitizing sensitive information, which can be deployed as a standalone service or docker container. Check out the [AI Sanitizer plugin](/hub/kong-inc/ai-sanitizer) configuration options for more information on how to configure the plugin with the AI PII Anonymizer Service.
```
The repo is private, so we can't link to it. I don't think this needs to be a note either - it's just standard info.

---
nav_title: Overview
---

> The Data Sanitization plugin for {{site.base_gateway}} helps protect sensitive information in client request bodies before they reach upstream services.

Author: ```suggestion
The AI Sanitizer plugin for {{site.base_gateway}} helps protect sensitive information in client request bodies before they reach upstream services.
```

---
nav_title: Overview
---

The AI Sanitizer plugin for {{site.base_gateway}} helps protect sensitive information in client request bodies before they reach upstream services.
By integrating with an external PII service, the plugin ensures compliance with data privacy regulations while preserving the usability of request data.
It supports multiple sanitization modes, including replacing sensitive information with fixed placeholders or generating synthetic replacements that retain category-specific characteristics.

> Additionally, it offers an optional restoration feature, allowing the original data to be reinstated in responses when needed.

Reviewer 2: ```suggestion
Additionally, AI Sanitizer offers an optional restoration feature, allowing the original data to be reinstated in responses when needed.
```

{:.note}
> This plugin extends the functionality of the [AI Proxy plugin](/hub/kong-inc/ai-proxy/) or the [AI Proxy Advanced](/hub/kong-inc/ai-proxy-advanced/) plugin, and requires an AI proxy to be configured first. 

> Check out the [AI Gateway quickstart](/gateway/latest/get-started/ai-gateway/) to get an AI proxy up and running within minutes!

Reviewer 2: ```suggestion
Check out the [AI Gateway quickstart](/gateway/latest/get-started/ai-gateway/) to get an AI proxy up and running within minutes.
```

> The AI Sanitizer plugin relies on the AI PII Anonymizer Service for identifying and sanitizing sensitive information, which can be deployed as a Docker container. See the [tutorial on configuring the AI Sanitizer plugin](/hub/kong-inc/ai-sanitizer/how-to/) for more information on how to configure the plugin with the AI PII Anonymizer Service.

Reviewer 2: ```suggestion
The AI Sanitizer plugin uses the AI PII Anonymizer Service, which can run in a Docker container, to detect and sanitize sensitive data.  See the [tutorial on configuring the AI Sanitizer plugin](/hub/kong-inc/ai-sanitizer/how-to/) for more information on how to configure the plugin with the AI PII Anonymizer Service.
```

## How it works

> 1. When a request reaches {{site.base_gateway}}, the plugin intercepts the request body and forwards it to the external PII service.
> 2. The PII service analyzes the content, identifies sensitive information, and applies the selected sanitization method (fixed placeholders or category-based synthetic replacements).
> 3. The sanitized request body is then forwarded to the upstream service through the AI Proxy or AI Proxy Advanced plugin.
> 4. If the restoration feature is enabled, the plugin can restore the original data in responses before returning them to the client, ensuring a seamless user experience.

Reviewer 2: ```suggestion
1. The plugin intercepts the request body and sends it to the external PII service.
1. The PII service detects sensitive data and applies the chosen sanitization method (placeholders or synthetic replacements).
1. The sanitized request is forwarded upstream with the AI Proxy or AI Proxy Advanced plugin.
1. If restoration is enabled, the plugin restores original data in responses before returning them to the client.
```

## AI PII Anonymizer service

Kong provides an [AI PII Anonymizer service](https://hub.docker.com/r/kong/ai-pii-service) Docker image. 

### Image configuration options

This service takes the following optional environment variables at startup:
* `GUNICORN_WORKERS`: Specifies the number of Gunicorn processes to run
* `PII_SERVICE_ENGINE_CONF`: Specifies the natural language processing (NLP) engine configuration file
* `GUNICORN_LOG_LEVEL`: Specifies log level

### Sanitization endpoints

* `POST /llm/v1/sanitize`: Sanitize specified types of PII information, including credentials, and custom patterns
* `POST /llm/v1/sanitize_credentials`: Only for sanitizing credentials

### Available anonymization modes

> You can anonymize data in requests using the following redact modes:
> 
> * `placeholder`: Redact the sensitive data with a fixed placeholder pattern, `PLACEHOLDER{i}`, where `i` is a sequence number and the same original text share the same number. 

Reviewer 2: ```suggestion
You can anonymize data in requests using the following redact modes:

* `placeholder`: Replaces sensitive data with a fixed placeholder pattern, `PLACEHOLDER{i}`, where `i` is a sequence number. Identical original values receive the same placeholder.
```

   
   For example, the location `New York City` might be replaced with `LOCATION`.
   
* `synthetic`: Redact the sensitive data with a word in the same type. 
   
   For example, the name `John` might be replaced with `Amir`.

>   * Custom patterns will be replaced with `CUSTOM{i}`.

Reviewer 2: ```suggestion
  * Custom patterns are replaced with `CUSTOM{i}`.
```

>   * For credentials, the string `#` with the same length as the original string will be used as a replacement.

Reviewer 2: ```suggestion
   * Credentials are replaced with a string of `#` characters matching the original length.
```

### Custom patterns

> You can introduce an array of custom patterns on a per-request basis. 
> Today we support regex patterns, and all fields are required (`name`, `regex`, and `score`).

Reviewer 2: ```suggestion
You can define an array of custom patterns on a per-request basis.  
Currently, only regex patterns are supported, and all fields are required: `name`, `regex`, and `score`.
```
